### DISCUSSION TITLE
RFC: New Task Filtering Syntax

### DISCUSSION THREAD

**Discussion by jaredpalmer** [vercel/turborepo#105] [RFC]
## Problem

The current filtering options in the CLI aren't very expressive. They are modeled after [Lerna's `--scope` flag](https://github.com/lerna/lerna/tree/main/core/filter-options#lernafilter-options). There are certain useful situations which are not really possible at the moment. One of them being executing things *up to but not including* a certain package task. 

## Impact / Context

This issue was raised by Lattice.com and Makeswift.com who both wanted more control over task filtering. 

## Solution

*Adopt [PNPM's filtering syntax](https://pnpm.io/filtering) for `turbo run` and make `turbo run` a pass through/default like `yarn <cmd>`.* Selectors may be specified via the `--filter` flag:

```bash
turbo <command> --filter <package_selector>
```

> The following is copied almost directly from PNPM...

### –filter <package_name>

To select an exact package, just specify its name (`@scope/pkg`) or use a pattern to select a set of packages (`@scope/*`).

Examples:

```bash
turbo test --filter "@babel/core"
turbo test --filter "@babel/*"
turbo test --filter "*core"
```

### –filter <package_name>…

To select a package and its dependencies (direct and non-direct), suffix the package name with an ellipsis: `<package_name>...`. For instance, the next command will run tests of `foo` and all of its dependencies:

```bash
turbo test --filter foo...
```

You may use a pattern to select a set of root packages:

```bash
turbo test --filter "@babel/preset-*..."
```

### –filter <package_name>^…

To ONLY select the dependencies of a package (both direct and non-direct), suffix the name with the aforementioned ellipsis preceded by a chevron. For instance, the next command will run tests for all of `foo`’s dependencies:

```bash
turbo test --filter "foo^..."
```

### –filter …<package_name>

To select a package and its dependent packages (direct and non-direct), prefix the package name with an ellipsis: `...<package_name>`. For instance, this will run the tests of `foo` and all packages dependent on it:

```bash
turbo test --filter ...foo
```

### –filter “…^<package_name>”

To ONLY select a package’s dependents (both direct and non-direct), prefix the package name with an ellipsis followed by a chevron. For instance, this will run tests for all packages dependent on `foo`:

```bash
turbo test --filter "...^foo"
```

### –filter ./<directory>

To only select packages under the specified directory, you may specify any relative path, typically in POSIX format.

### –filter {<directory>}

Includes all projects that are under the specified directory.

It may be used with the ellipsis and chevron operators to select dependents/dependencies as well:

```bash
turbo <cmd> --filter ...{<directory>}
turbo <cmd> --filter {<directory>}...
turbo <cmd> --filter ...{<directory>}...
```

It may also be combined with `[<since>]`. For instance, to select all changed projects inside a directory:

```bash
turbo <cmd> --filter "{packages}[origin/master]"
turbo <cmd> --filter "...{packages}[origin/master]"
turbo <cmd> --filter "{packages}[origin/master]..."
turbo <cmd> --filter "...{packages}[origin/master]..."
```

Or you may select all packages from a directory with names matching the given pattern:

```
turbo <cmd> --filter "@babel/*{components}"
turbo <cmd> --filter "@babel/*{components}[origin/master]"
turbo <cmd> --filter "...@babel/*{components}[origin/master]"
```

### –filter “[<since>]”

Selects all the packages changed since the specified commit/branch. May be suffixed or prefixed with `...` to include dependencies/dependents.

For example, the next command will run tests in all changed packages since `master` and on any dependent packages:

```
turbo test --filter "...[origin/master]"
```

### Multiplicity

When packages are filtered, every package is taken that matches at least one of the selectors. You can use as many filters as you want:

```bash
turbo test --filter ...foo --filter bar --filter baz...
```

### Excluding

Any of the filter selectors may work as exclusion operators when they have a leading “!”. In zsh (and possibly other shells), “!” should be escaped: `\!`.

For instance, this will run a command in all projects except for `foo`:

```bash
turbo <cmd> --filter=!foo
```

And this will run a command in all projects that are not under the `lib` directory:

```bash
turbo <cmd> --filter=!./lib
```

### –filter-prod <filtering_pattern>

Acts the same a `--filter` but omits `devDependencies` when selecting dependency projects from the workspace.

### –test-pattern <glob>

`test-pattern` allows detecting whether the modified files are related to tests. If they are, the dependent packages of such modified packages are not included.

This option is useful with the “changed since” filter. For instance, the next command will run tests in all changed packages, and if the changes are in the source code of the package, tests will run in the dependent packages as well:

```
turbo --filter="...[origin/master]" --test-pattern="test/*" test
```

### –changed-files-ignore-pattern <glob>

Allows to ignore changed files by glob patterns when filtering for changed projects since the specified commit/branch.

Usage example:

```bash
turbo --filter="...[origin/master]" --changed-files-ignore-pattern="**/README.md" run build
```

## Discussion

- One thing I (@Jared) don’t care for in the above is the meaning of the `...` . I actually think PNPM (which the above knocks off) has it *exactly* backwards. For example, to `build` in a  package and its own dependencies (that is the things listed it its own `package.json`, you run this in pnpm: `pnpm build --filter=front...`. I
    
    I find this pretty awkward and hard to remember. I think preceding `...` should be *dependencies* and `...` after should be *dependents*. To me this would make it closer to the English usage of ellipses: 
    
    - Before → "up to it" = *it plus its dependencies.*
    - After → "it and everything it impacts" *= it plus its dependents*
    
    Unfortunately, this is exactly backwards in PNPM. 
    
- Turbo can run multiple commands at the same time. I wonder if we tweaked the syntax to combine the build target and the filter into one single block:
    
    ```bash
    turbo ...{./packages/*}::build[origin/master]    ...^::thing#test
    ```
    
    We would need to alter the DSL to remove ambiguity. 
    
    - `::` would delimit an npm `package.json` script. However, it should required for disambiguation so that `turbo build` and `turbo test` etc. are the same as `turbo ::build` and `turbo ::test`
    - turbo currently uses `^` in the pipeline definition to express a topological task relationship. We should switch this to `...` in order to make the filtering syntax and pipeline congruent.
        
        ```json
        { 
          "turbo": {
            "pipeline": {
              "build": { 
                "dependsOn": ["^build"] // this should be ... 
              }, 
            }
          }
        }
        ```
        
    - turbo currently uses `#` as a delimiter (e.g.  `<package>#<npm-pkgjson-script>` ) in pipeline definition for package task-specific overrides. We should switch to `::` for congruency with filtering.
        
        ```json
        { 
          "turbo": {
            "pipeline": {
              "build": { 
                "dependsOn": ["^build"],
                "outputs": ["dist/**"]
              }, 
              "front#build": { // this should become :: to match CLI filter
                "dependsOn": ["^build"],
                "outputs": ["dist/**"]
              }, 
            }
          }
        }
        ```
        
        *Note: the reason we can’t use  `.`  as the delimiter is because a  `.`  is allowed in NPM package names (e.g. `lodash.mapkeys` is a valid package).*

**Comment by paul-sachs** [comment]
I've also noticed that the current --scope option behaves a little weird. I have a few applications and one app depends on another (AppA depends on AppB) and if I run `turbo run start --scope="AppB"` it seems to run both AppA and AppB. Ignore does nothing to affect the run. Not sure if this is a bug or what.

**Comment by ekosz** [comment]
As one of the original requestors let me try and argue why the pnpm `...` makes a lot of sense as it its. Lets say we have the dependency chain: `A->B->C->D->E`

If I write `pnpm run lint --filter="...C". I know it's going to run "A, B, C" because I look to the left of "C" in the dependency chain. Same for "C..." I just look to the right. The dots match up with the left and right of the chain.

**Comment by NullVoxPopuli** [comment]
Would this solve? https://github.com/vercel/turborepo/issues/417
all the examples have the same command running in all packages.
tldr: I want to run "start" in some target package, and "build" in all dependencies (before running "start" in the target package)

**Reply by gsoltis** [reply]
Unfortunately, not yet. There is one further feature required, which is splitting entrypoint selection (your app) from the resolution of pipeline dependencies (build all of its dependencies). Currently, unless your app's dependencies are also selected by the combination of `scope`/`filter`/`since` etc., they won't be included in the pipeline dependency calculation.

**Reply by gsoltis** [reply]
I filed #912 to track the issue

**Comment by kaminskypavel** [comment]
actually working pretty great in canary.4, thanks @gsoltis .

added to my package.json
```
    "build:packages": "turbo run build --filter={packages/*}",
    "build:apps": "turbo run build --filter={apps/*}",
```

**Comment by wilsonlewis** [comment]
How about tagging system in the `package.json`? Sort of similar to this example:

https://youtu.be/LUSPTD0Ithc?t=1085

```
// PackageA/package.json
{
  tags: ['feature-a']
}
```

```
// PackageB/package.json
{
  tags: ['feature-a', 'feature-b']
}
```

```
// PackageC/package.json
{
  tags: ['feature-b', 'feature-c']
}
```

```
turbo run dev --tags feature-a,feature-b
```

**Reply by theoephraim** [reply]
huge +1 for a more flexible tagging setup. In my case I'd end up marking various packages with parts of the stack (ex: `frontend` vs `backend`) or the technologies they use (ex: `ts`, `rust`, etc). It would allow me to keep my task naming simple, but still filter the packages easily.

This coupled with using filters in pipeline/dependencies would be huge.

cc @jaredpalmer

**Comment by jaredpalmer** [comment]
The new `--filter` flag is now available in `turbo@1.2+`!

See the documentation here: https://turborepo.org/docs/features/filtering
